### PR TITLE
Added document about live streaming

### PR DESCRIPTION
**Pull request by ethouris**

### CODE REVIEW

File: docs/live-streaming.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+MPEG-TS is the most important protocol predicted to be sent over internet using
+SRT and the main reason of initiating this project. SRT obviously isn't limited
+to MPEG-TS, howver it still is limited in use to any data transmission that can
+be called Live Streaming - at least if you use the Live mode (which is SRT
+default).
+
+MPEG-TS consists of single units of 188 bytes. By having `188*7` we get the
+size of 1316, which is the maximum multiplicity of 188 that results still less
+than 1500, the standard MTU size in UDP transport, which is the underlying
```
Comment (adamyellen) [inline docs/live-streaming.md]: standard Ethernet MTU size

File: docs/live-streaming.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+the same AVERAGE speed, as the video player that would play it, or even more
+exactly, the data must be produced exactly the same fast as they will be consumed
+by the video player.
+
+
+Live Streaming Process
+======================
+
+Now that you know how the Live Streaming should turn a bunch of MPEG-TS
+encoded video stream into a network live stream, let's complete the definition
```
Comment (adamyellen) [inline docs/live-streaming.md]: encoded video frames?

Comment (ethouris) [reply docs/live-streaming.md]: I'd even add "encoded video and audio frames", although I'm not sure if this isn't too elaborate.

File: docs/live-streaming.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+MPEG-TS is the most important protocol predicted to be sent over internet using
+SRT and the main reason of initiating this project. SRT obviously isn't limited
+to MPEG-TS, howver it still is limited in use to any data transmission that can
+be called Live Streaming - at least if you use the Live mode (which is SRT
+default).
+
+MPEG-TS consists of single units of 188 bytes. By having `188*7` we get the
+size of 1316, which is the maximum multiplicity of 188 that results still less
+than 1500, the standard MTU size in UDP transport, which is the underlying
+transport layer for SRT. The headers for Ethernet, IP and UDP protocol occupy
```
Comment (StephaneStH) [inline docs/live-streaming.md]: 28 is only IP header (20) + UDP (8), ethernet headers are not part of IP (not UDP) MTU

File: docs/live-streaming.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+Transmitting MPEG TS binary protocol over SRT
+=============================================
+
+MPEG-TS is the most important protocol predicted to be sent over internet using
+SRT and the main reason of initiating this project. SRT obviously isn't limited
+to MPEG-TS, howver it still is limited in use to any data transmission that can
+be called Live Streaming - at least if you use the Live mode (which is SRT
+default).
+
+MPEG-TS consists of single units of 188 bytes. By having `188*7` we get the
```
Comment (StephaneStH) [inline docs/live-streaming.md]: "individual cells" could be an improvement instead of "single units".

File: docs/live-streaming.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+fulfills only and exclusively the transmission part. Live Streaming process
+consists of more parts, though.
+
+
+Transmitting MPEG TS binary protocol over SRT
+=============================================
+
+MPEG-TS is the most important protocol predicted to be sent over internet using
+SRT and the main reason of initiating this project. SRT obviously isn't limited
+to MPEG-TS, howver it still is limited in use to any data transmission that can
```
Comment (StephaneStH) [inline docs/live-streaming.md]: Typo, should be 'however'.

File: docs/live-streaming.md (hunk 3 above)
Comment (StephaneStH) [reply docs/live-streaming.md]: "IP (not UDP) MTU"... actually ethernet MTU like Adam said.

Comment (ethouris) [reply docs/live-streaming.md]: Ok, Ethernet removed from this "list".

File: docs/live-streaming.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+which come from two reasons:
+
+1. Sending over the network has only average time perdictability. Unusual and
+unexpected delays, coming from both the network and your system, may slightly
+destroy your perfect time calculations. Therefore you should not rely on the
+exact number of bytes sent by one "frame package", but rather on overall
+transmission size, and do often periodic synchronization of the transmission
+time basing on the timestamps in the stream.
+
+2. Not all frames as I-Frames. Most of the frames sent in the video stream are
```
Comment (rndi) [inline docs/live-streaming.md]: Not all frames *are*

File: docs/live-streaming.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+tolerance, which results from size difference. what must be absolutely adhered
+to, however, is that the distance between the first network unit transporting
+the first portion of one I-Frame, and the same unit for the next I-Frame must
+correspond with the time distance between these two same I-frames that results
+from their timestamps.
+
+In other words: there is some slight buffering at the receiving side, but the
+requirement for a live streaming is that: data must be transmitted with exactly
+the same AVERAGE speed, as the video player that would play it, or even more
+exactly, the data must be produced exactly the same fast as they will be consumed
```
Comment (rndi) [inline docs/live-streaming.md]: produced at exactly the same speed as it will be

File: docs/live-streaming.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+======================
+
+Now that you know how the Live Streaming should turn a bunch of MPEG-TS
+encoded video stream into a network live stream, let's complete the definition
+of the live streaming transport over the network using SRT.
+
+The source side should be a real live stream, that is, for example:
+
+- a file read by an application that can make a live stream out of it
+- a grab device that is grabbing the data in constant rates and then
```
Comment (rndi) [inline docs/live-streaming.md]: a capture device that is capturing data at constant rates and passing the data to to a live encoder

File: docs/live-streaming.md [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+the data are passed through the live encoder
+- some existing live network encoder that streams a live stream over UDP
+- a camera that uses some simple encoding method, like MJPEG (needs transcoding)
```
Comment (rndi) [inline docs/live-streaming.md]: remove "needs transcoding"

File: docs/live-streaming.md [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+can be `ffmpeg`. You just need to remember about two things here:
+
+- The `-re` options is required for making the live stream out of a file
+- The `pkt_size=1316` parameter should be added to UDP output URI, if you make
+ffmpeg output to UDP
+
+Pay attention first that the splitting the data for transport into single units
+that can fit in one UDP packet, and first of all defining time distances
+between these units is something that must come in synch with the timestamps
+in the transport stream. Even if you have a grab device, it usually will make
```
Comment (rndi) [inline docs/live-streaming.md]: capture device, it will usually capture

File: docs/live-streaming.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+receiving side before it's delivered to the output. This time should cover
+both the unexpectedly delayed transmission of one UDP packet, as well
+as have extra time for a case when the packet was lost and had to be
+retransmitted. Every UDP packet carrying an SRT packet has a timestamp,
+which is grabbed at the time when the packet is given up to SRT for
+sending, and using that timestamp the appropriate delay is applied before
+delivering to the output, so that time distances between two consecutive
+packets at the delivery application is aimed to be exactly the same as
+the time distance between these same packets at the time when they are
+given up to SRT for sending.
```
Comment (rndi) [inline docs/live-streaming.md]: given to SRT

File: docs/live-streaming.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+28 bytes in this size, leaving 1472 bytes of size, and SRT occupies next 16
+bytes for its own header, which leaves the maximum payload size 1456 bytes. The
+MPEG-TS transport unit size still fits in it, and fortuantely this is a
+standard unit size used in Live Streaming, also with the use of UDP.
+
+SRT isn't limited to MPEG-TS - you can use any other data format that you
+think can be suitable, as well as use some your own higher level protocol
+on top of MPEG-TS and have your extra header (you still have extra 140 bytes
+per unit for that purpose). But even though that, the transmission must still
+satisfy the Live Streaming Requirements.
```
[the remark this answers is not in the document]
Comment (ethouris) [reply docs/live-streaming.md]: No. This is an extra information that states how SRT isn't tightly bound to MPEG-TS, but it's stream type agnostic.

File: docs/live-streaming.md (hunk 7 above)
Comment (ethouris) [reply docs/live-streaming.md]: These are synonyms, and the term of "the same speed" is used in this sentence before. For the second part of the sentence this shouldn't be a big problem to use a small abbreviation. Using shorter expression with direct understanding improves comprehension, even if they sometimes sound too informal.

File: docs/live-streaming.md (hunk 8 above)
Comment (ethouris) [reply docs/live-streaming.md]: I'd use rather "a grab device that is capturing data" (the "grab device" term is taken from the ffmpeg documentation).

File: docs/live-streaming.md (hunk 11 above)
Comment (ethouris) [reply docs/live-streaming.md]: passed to SRT should sound better anyway

File: docs/live-streaming.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+default).
+
+MPEG-TS consists of single units of 188 bytes. By having `188*7` we get the
+size of 1316, which is the maximum multiplicity of 188 that results still less
+than 1500, the standard MTU size in UDP transport, which is the underlying
+transport layer for SRT. The headers for Ethernet, IP and UDP protocol occupy
+28 bytes in this size, leaving 1472 bytes of size, and SRT occupies next 16
+bytes for its own header, which leaves the maximum payload size 1456 bytes. The
+MPEG-TS transport unit size still fits in it, and fortuantely this is a
+standard unit size used in Live Streaming, also with the use of UDP.
```
[the remark this answers is not in the document]
Comment (ethouris) [reply docs/live-streaming.md]: The sentence was slightly reformed.

**Pull request merged**